server/src/game/client_actor.rs

```diff
@@ -0,0 +1,125 @@
+        Self {
+            game_name: None,
+            game_server,
+            last_heartbeat: Instant::now(),
+            session_info: session,
+        }
+    }
+
+    /// Begins sending heartbeats to the user.
+    fn start_heartbeats(&self, ctx: &mut ws::WebsocketContext<Self>) {
+        ctx.run_interval(HEARTBEAT_INTERVAL, move |client, ctx| {
+            if Instant::now().duration_since(client.last_heartbeat) > CLIENT_TIMEOUT {
+                println!("Heartbeat timed out, disconnecting.");
+                client.game_server.do_send(Disconnect {
+                    user_id: client.session_info["user_id"].clone(),
+                    game_name: client.game_name.clone(),
+                });
+                ctx.stop();
+                return;
+            }
+
+            ctx.ping(b"Ping!");
+        });
+    }
+}
+
+impl Actor for WsClient {
+    type Context = ws::WebsocketContext<Self>;
+
+    fn started(&mut self, ctx: &mut Self::Context) {
+        self.start_heartbeats(ctx);
+
+        let address = ctx.address();
+
+        self.game_server
+            .send(Connect {
+                address: address.recipient(),
+                user_id: self.session_info["user_id"].clone(),
+            })
+            .into_actor(self)
+            .then(|result, _client, ctx| {
+                match result {
+                    Ok(_) => (),
+                    _ => ctx.stop(),
+                }
+                fut::ready(())
+            })
+            .wait(ctx);
+    }
+
+    fn stopping(&mut self, _ctx: &mut Self::Context) -> Running {
+        self.game_server.do_send(Disconnect {
+            user_id: self.session_info["user_id"].clone(),
+            game_name: self.game_name.clone(),
+        });
+        Running::Stop
+    }
+}
+
+/// Handle messages from the server
+impl Handler<ServerMessage> for WsClient {
+    type Result = ();
+
+    fn handle(&mut self, _msg: ServerMessage, _ctx: &mut Self::Context) {
+        todo!()
+    }
+}
+
+/// Handle messages from the user
+impl StreamHandler<Result<ws::Message, ws::ProtocolError>> for WsClient {
+    fn handle(&mut self, msg: Result<ws::Message, ws::ProtocolError>, ctx: &mut Self::Context) {
+        let msg = match msg {
+            Err(_) => {
+                ctx.stop();
+                return;
+            }
+            Ok(msg) => msg,
+        };
+
+        match msg {
+            ws::Message::Text(_text) => {
+                // "Text" type messages are successful game messages
+                // TODO: all of the user-client message handling :)
+                todo!()
+            }
+            ws::Message::Ping(msg) => {
+                self.last_heartbeat = Instant::now();
+                ctx.pong(&msg);
+            }
+            ws::Message::Pong(_) => {
+                self.last_heartbeat = Instant::now();
+            }
+            ws::Message::Binary(_) => panic!("Unexpected binary"),
```

Comment:
Is it safe to panic when we receive binary from a client? If a client panicking propagates into the actix server worker panicking, then I think this is bad behavior. Clients should not be able to cause server workers to panic, we should probably just log an error here and move on.